Incorrect Comparable Expression Produced by Ternary Expression

# CFR version
CFR 0.153-SNAPSHOT (24c7433-dirty)


# Compiler
javac 11.0.20.1


# Description
The ternary expression may be decompiled into an incorrect Comparable expression.


# Example
The example code is as follows:

```java
class T1 {
    public Object foo() {
        Boolean b1 = false;
        Boolean b2 = true;
        final Object o = ((b1) ? -33.998 : b2);
        return o;
    }
}
```

The decompiled code is as follows:

```java
/*
 * Decompiled with CFR 0.153-SNAPSHOT (24c7433-dirty).
 */
class T1 {
    T1() {
    }

    public Object foo() {
        Boolean bl = false;
        Boolean bl2 = true;
        Comparable<Boolean> comparable = bl != false ? (Comparable<Double>)Double.valueOf(-33.998) : (Comparable<Double>)bl2;
        return comparable;
    }
}
```

Compiling the code results in the following compilation error:

```txt
./T1.java:11: error: incompatible types: bad type in conditional expression
        Comparable<Boolean> comparable = bl != false ? (Comparable<Double>)Double.valueOf(-33.998) : (Comparable<Double>)bl2;
                                                       ^
    Comparable<Double> cannot be converted to Comparable<Boolean>
./T1.java:11: error: incompatible types: Boolean cannot be converted to Comparable<Double>
        Comparable<Boolean> comparable = bl != false ? (Comparable<Double>)Double.valueOf(-33.998) : (Comparable<Double>)bl2;
                                                                                                                         ^
2 errors
```

--- a/src/org/benf/cfr/reader/bytecode/analysis/types/discovery/InferredJavaType.java
+++ b/src/org/benf/cfr/reader/bytecode/analysis/types/discovery/InferredJavaType.java
@@ -149,7 +149,22 @@ public class InferredJavaType {
                     continue;
                 }
                 Map<? extends JavaTypeInstance, JavaGenericRefTypeInstance> boundSupers = otherSupers.getBoundSuperClasses();
-                matches.keySet().retainAll(boundSupers.keySet());
+
+                Iterator<Map.Entry<JavaTypeInstance, JavaGenericRefTypeInstance>> matchesIter = matches.entrySet().iterator();
+                while (matchesIter.hasNext()) {
+                    Map.Entry<JavaTypeInstance, JavaGenericRefTypeInstance> entry = matchesIter.next();
+                    if (!boundSupers.containsKey(entry.getKey())) {
+                        matchesIter.remove();
+                    } else {
+                        JavaGenericRefTypeInstance matchGeneric = entry.getValue();
+                        JavaGenericRefTypeInstance otherGeneric = boundSupers.get(entry.getKey());
+
+                        if (!MiscUtils.objectsEquals(matchGeneric, otherGeneric)) {
+                            // TODO: Try to determine common generic supertype?
+                            matchesIter.remove();
+                        }
+                    }
+                }
             }
             return matches;
         }

--- a/src/org/benf/cfr/reader/util/MiscUtils.java
+++ b/src/org/benf/cfr/reader/util/MiscUtils.java
@@ -5,6 +5,7 @@ import org.benf.cfr.reader.bytecode.analysis.parse.LValue;
 import org.benf.cfr.reader.bytecode.analysis.parse.expression.LValueExpression;
 import org.benf.cfr.reader.bytecode.analysis.parse.lvalue.LocalVariable;
 import org.benf.cfr.reader.bytecode.analysis.types.JavaTypeInstance;
+import org.benf.cfr.reader.util.annotation.Nullable;
 import org.benf.cfr.reader.util.functors.Predicate;
 
 import java.util.regex.Pattern;
@@ -46,6 +47,10 @@ public class MiscUtils {
      * up by letting code analysers say we're no good kids, hanging around wasting time.
      */
     public static void handyBreakPoint() {
+    }
+
+    public static boolean objectsEquals(@Nullable Object o1, @Nullable Object o2) {
+        return o1 == o2 || o1 != null && o1.equals(o2);
     }
 
     public static boolean isThis(Expression obj, JavaTypeInstance thisType) {

